### PR TITLE
CP-47358, CP-47361: Generate convert functions Go code

### PR DESCRIPTION
**Pull request by duobei**
1. Add templates for convert functions
2. Generate convert functions Go code

### CODE REVIEW

File: ocaml/sdk-gen/go/gen_go_binding.ml [hunk 1]
```diff
@@ -44,10 +44,59 @@ let render_api_messages_and_errors destdir =
...
+  let event = render_template "ConvertBatch.mustache" Convert.event_batch () in
+  let interface =
+    render_template "ConvertInterface.mustache" Convert.interface ()
+  in
+  let param_types = Types.of_params objects in
+  let result_types = Types.of_results objects in
+  let generate types of_json =
+    types
+    |> List.map Convert.of_ty
+    |> List.map (fun params ->
```
Comment (contificate) [inline ocaml/sdk-gen/go/gen_go_binding.ml:76]: Can probably perform fusion here to remove the intermediary list.

In the pure setting, `map g (map f xs) = map (g . f) xs` always holds (saving you the redundant intermediary list); in this case, I suspect there is no sequencing of side effects that we must respect (tell me if I'm wrong, as I don't know the semantics of `Convert`!). 

A minor thing, but probably want to push it inwards as the compiler will likely make no effort to:
```ocaml
    types
    |> List.map (fun params ->
           let template = Convert.(template_of_convert (of_ty params)) in
```

File: ocaml/sdk-gen/go/gen_go_helper.ml [hunk 2]
```diff
@@ -198,3 +233,309 @@ module Json = struct
   let api_errors =
     List.map (fun error -> `O [("name", `String error)]) !Api_errors.errors
 end
+
+module Types = struct
```
Comment (minglumlu) [inline ocaml/sdk-gen/go/gen_go_helper.ml:237]: I think this module could be in `CommonFunctions.ml`.

File: ocaml/sdk-gen/go/gen_go_helper.ml [hunk 3]
```diff
@@ -198,3 +233,309 @@ module Json = struct
...
+    let return_types =
+      let aux accu msg =
+        match msg.msg_result with None -> accu | Some (ty, _) -> ty :: accu
+      in
+      mesages objects |> List.fold_left aux [] |> Listext.List.setify
+    in
+    List.concat_map decompose return_types |> Listext.List.setify
+end
+
+module Convert = struct
```
Comment (minglumlu) [inline ocaml/sdk-gen/go/gen_go_helper.ml:280]: I'm happy with this module now unless other reviewers have other thoughts.
Minor comment on the "func_partial_type"/"fp_type". The naming looks not great. This was also raised in another PR.

File: ocaml/sdk-gen/go/gen_go_helper.mli [hunk 4]
```diff
@@ -30,3 +30,74 @@ module Json : sig
 
   val api_errors : Mustache.Json.value list
 end
+
+module Types : sig
```
Comment (minglumlu) [inline ocaml/sdk-gen/go/gen_go_helper.mli:34]: The module name may better be `TypesOfMessages`

Comment (duobei) [reply ocaml/sdk-gen/go/gen_go_helper.mli]: Renamed to `TypesOfMessages`.

File: ocaml/sdk-gen/go/gen_go_helper.ml (hunk 2 above)
Comment (duobei) [reply ocaml/sdk-gen/go/gen_go_helper.ml]: I have moved it to `CommonFunctions.ml`.

File: ocaml/sdk-gen/go/gen_go_binding.ml (hunk 1 above)
Comment (duobei) [reply ocaml/sdk-gen/go/gen_go_binding.ml]: Nice. I adopted this.

**Comment by xueqingz** [comment]
Checking the Go lint format pass with:
```shell
root@3eea0d3e0e5d:/app/go/goSDK# golangci-lint run --config=/app/.golangci.yml
WARN The linter 'gomnd' is deprecated (since v1.58.0) due to: The linter has been renamed. Replaced by mnd. 
WARN The linter 'execinquery' is deprecated (since v1.58.0) due to: The repository of the linter has been archived by the owner.  
0 issues.
``` 

Test with xenserver-samples Go project. No special issues found on 7b88aa5089bd7df8d56d40fd471726e19a86ba11.

**Comment by xueqingz** [comment]
The test Go files naming with different format , could we use same rules for all convert test go files? like xx_convert.go?

File: ocaml/sdk-gen/go/test_data/record_convert.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	return
+}
+
+func deserializeVBDRecord(context string, input interface{}) (record VBDRecord, err error) {
+	rpcStruct, ok := input.(map[string]interface{})
+	if !ok {
+		err = fmt.Errorf("failed to parse XenAPI response: expected Go type %s at %s but got Go type %s with value %v", "map[string]interface{}", context, reflect.TypeOf(input), input)
+		return
+	}
+	uuidValue, ok := rpcStruct["uuid"]
```
Comment (xueqingz) [inline ocaml/sdk-gen/go/test_data/record_convert.go:20]: The test result here is different with generated Go method:
```go
uUIDValue, ok := rpcStruct["uuid"]
	if ok && uUIDValue != nil {
		record.UUID, err = deserializeString(fmt.Sprintf("%s.%s", context, "uuid"), uUIDValue)
		if err != nil {
			return
		}
	}
``` 
uuidValue , uUIDValue , why ?

File: ocaml/sdk-gen/go/test_data/batch.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+		}
+	}
+	validRefCountsValue, ok := rpcStruct["valid_ref_counts"]
+	if ok && validRefCountsValue != nil {
+		batch.ValidRefCounts, err = deserializeStringToIntMap(fmt.Sprintf("%s.%s", context, "valid_ref_counts"), validRefCountsValue)
+		if err != nil {
+			return
+		}
+	}
+	return
```
Comment (xueqingz) [inline ocaml/sdk-gen/go/test_data/batch.go:21]: Could you add all parameters for Event Batch ? it's a special convert , we'd want to test all.
In generated file, it's:
```go
func deserializeEventBatch(context string, input interface{}) (batch EventBatch, err error) {
	rpcStruct, ok := input.(map[string]interface{})
	if !ok {
		err = fmt.Errorf("failed to parse XenAPI response: expected Go type %s at %s but got Go type %s with value %v", "map[string]interface{}", context, reflect.TypeOf(input), input)
		return
	}
	tokenValue, ok := rpcStruct["token"]
	if ok && tokenValue != nil {
		batch.Token, err = deserializeString(fmt.Sprintf("%s.%s", context, "token"), tokenValue)
		if err != nil {
			return
		}
	}
	validRefCountsValue, ok := rpcStruct["validRefCounts"]
	if ok && validRefCountsValue != nil {
		batch.ValidRefCounts, err = deserializeStringToIntMap(fmt.Sprintf("%s.%s", context, "validRefCounts"), validRefCountsValue)
		if err != nil {
			return
		}
	}
	eventsValue, ok := rpcStruct["events"]
	if ok && eventsValue != nil {
		batch.Events, err = deserializeEventRecordSet(fmt.Sprintf("%s.%s", context, "events"), eventsValue)
		if err != nil {
			return
		}
	}
	return
}
```

Comment (duobei) [reply ocaml/sdk-gen/go/test_data/batch.go:21]: Yeah, I will add all.

File: ocaml/sdk-gen/go/gen_go_binding.ml [hunk 7]
```diff
@@ -44,10 +44,59 @@ let render_api_messages_and_errors destdir =
...
+           let json : Mustache.Json.t = of_json params in
+           render_template template json ()
+       )
+    |> String.concat ""
+  in
+  let rendered =
+    let serializes_rendered = generate param_types Convert.of_serialize in
+    let deserializes_rendered = generate result_types Convert.of_deserialize in
+    render_convert_header ()
+    ^ serializes_rendered
```
[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/sdk-gen/go/gen_go_binding.ml:111]: Sorry, I had forgotten to submit this comment a while ago when I made it. This is just a minor point

[the remark this answers is not in the document]
Comment (duobei) [reply ocaml/sdk-gen/go/gen_go_binding.ml:111]: > `^` is expensive; better to use `String.concat` if this code is run often.

Yeah, I have refactored the code using `String.concat`. Thanks.

[the remark this answers is not in the document]
Comment (duobei) [reply ocaml/sdk-gen/go/gen_go_binding.ml:111]: > Sorry, I had forgotten to submit this comment a while ago when I made it. This is just a minor point

Thanks all the same.

File: ocaml/sdk-gen/go/test_data/record_convert.go (hunk 5 above)
Comment (duobei) [reply ocaml/sdk-gen/go/test_data/record_convert.go:20]: `uuidValue  `is a sample data for testing template. `uUIDValue` is real generated data. It's a little weird now. I'll fixed this later.

**Review by xueqingz — APPROVED** [review]
Golang lint check pass on commit 663e38, no special error found during samples test.

**Pull request merged**